### PR TITLE
(Chore) Fix Tests + Improve Tooling for Testing

### PR DESCRIPTION
**Pull request by mrosnerr**
# Summary
Fix tests and add tooling to help promote them being maintained going forward.

# Details
The changes here:
1. Fix failing tests
2. Added a script to support `yarn test`
3. Added test running to `lint-staged` to tie it into pre-commit hook checks
4. Added a github action to validate PR's which includes linting, testing, and building ([test run](https://github.com/mrosnerr/nestjs-trpc/actions/runs/12216936478/job/34080569157) (see call out re. failure below))

## Call outs

1. The only test that still fails is [router.decorator.spec.ts](https://github.com/KevinEdry/nestjs-trpc/blob/c4e45b453c47f526b2b8e094b77feac9a21e42dc/packages/nestjs-trpc/lib/decorators/__tests__/router.decorator.spec.ts) which requires the change from #47 merged to pass (tested locally)
2. I made my own comments on the code with things I though would be noteworthy on top of conforming tests to the current functionality

Happy to make any adjustments or talk through the changes below - cheers 🍻

### CODE REVIEW

File: packages/nestjs-trpc/lib/decorators/__tests__/middlewares.decorator.spec.ts [hunk 1]
```diff
@@ -4,24 +4,28 @@ import { MIDDLEWARES_KEY } from '../../trpc.constants';
 import { MiddlewareOptions, MiddlewareResponse, TRPCMiddleware } from '../../interfaces';
 
 describe('UseMiddlewares Decorator', () => {
-  const mockMiddleware = () => ({});
+  class TestMiddleware implements TRPCMiddleware {
+    use(opts: MiddlewareOptions<object>): MiddlewareResponse | Promise<MiddlewareResponse> {
+      throw new Error('Method not implemented.');
+    }
+  }
```
Comment (mrosnerr) [inline packages/nestjs-trpc/lib/decorators/__tests__/middlewares.decorator.spec.ts:11]: [note] The typing enforces a class now so I moved all the tests to use this convention

Feel free to resolve

File: packages/nestjs-trpc/lib/factories/middleware.factory.ts [hunk 2]
```diff
@@ -30,7 +30,7 @@ export class MiddlewareFactory {
       );
 
       const procedureMiddleware = procedures.flatMap((procedure) => {
-        return procedure.middlewares != null ? procedure.middlewares : [];
+        return procedure.middlewares ?? [];
```
Comment (mrosnerr) [inline packages/nestjs-trpc/lib/factories/middleware.factory.ts:33]: [note] this was coming through as `undefined` in one of the tests so I changed it to handle either.

Feel free to resolve

File: packages/nestjs-trpc/lib/generators/__tests__/router.generator.spec.ts [hunk 3]
```diff
@@ -152,6 +166,9 @@ describe('RouterGenerator', () => {
         },
       ];
 
+      procedureGenerator.generateProcedureString.mockReturnValueOnce('testQuery: publicProcedure.query(async () => "PLACEHOLDER_DO_NOT_REMOVE" as any )');
+      procedureGenerator.generateProcedureString.mockReturnValueOnce('testMutation: publicProcedure.mutation(async () => "PLACEHOLDER_DO_NOT_REMOVE" as any )');
```
Comment (mrosnerr) [inline packages/nestjs-trpc/lib/generators/__tests__/router.generator.spec.ts:170]: [note] The actual generation of these seemed important so I moved it to it's own test: [procedure.generator.spec.ts](https://github.com/KevinEdry/nestjs-trpc/pull/49/files#diff-65dcad7e46147205bd366f0401ed4ce9056b04f5afcb5c67833ab71baacbe791)

Feel free to resolve

**Comment by KevinEdry** [comment]
This is amazing! I wanted to make this update for such a long time now but couldn't find the time.

**Pull request merged**